docs/database_migration.md

```diff
@@ -6,52 +6,87 @@ NOTE: [pgloader](https://github.com/dimitri/pgloader) is a tool that can easily
+4. Insert data into target
+5. Change LLDAP config to new target and restart
 
 The steps below assume you already have PostgreSQL or MySQL set up with an empty database for LLDAP to use.
 
-## Create a dump
+## Create schema on target
 
-First, we must dump the existing data to a file. The dump must be tweaked slightly according to your target db. See below for commands
+LLDAP has a command that will connect to a target database and initialize the
+schema. If running with docker, run the following command to use your active
+instance (this has the benefit of ensuring your container has access):
 
-### PostgreSQL
+```
+docker exec -it <LLDAP container name> /app/lldap create_schema -d <Target database url>
+```
+
+If it succeeds, you can proceed to the next step.
 
-PostgreSQL uses a different hex string format and doesn't support `PRAGMA`.
+## Create a dump of existing data
+
+We want to dump (almost) all existing values to some file - the exception being the `metadata` table. Be sure to stop/pause LLDAP during this step, as some
+databases (SQLite in this example) will give an error if LLDAP is in the middle of a write. The dump should consist just INSERT
+statements. There are various ways to do this, but a simple enough way is filtering a
+whole database dump. For example:
 
 ```
-sqlite3 /path/to/lldap/config/users.db .dump | \
-sed -r -e "s/X'([[:xdigit:]]+'[^'])/'\\\x\\1/g" \
--e 's/^CREATE TABLE IF NOT EXISTS "([^"]*)".*/DELETE FROM \1;/' \
--e '/^PRAGMA.*/d' > /path/to/dump.sql
+sqlite3 /path/to/lldap/config/users.db .dump | grep "^INSERT" | grep -v "^INSERT INTO metadata" > /path/to/dump.sql
 ```
 
-### MySQL
+## Sanitize data
+
+Some databases might use different formats for some data - for example, PostgreSQL uses
+a different syntax for hex strings than SQLite. We also want to make sure inserts are done in
+a transaction in case one of the statements fail.
+
+### To PostgreSQL
 
-MySQL doesn't support `PRAGMA`.
+PostgreSQL uses a different hex string format. The command below should switch SQLite
+format to PostgreSQL format, and wrap it all in a transaction:
 
 ```
-sqlite3 /path/to/lldap/config/users.db .dump | \
--e 's/^CREATE TABLE IF NOT EXISTS "([^"]*)".*/DELETE FROM \1;/' \
--e '/^PRAGMA.*/d' > /path/to/dump.sql
+sed -i -r -e "s/X'([[:xdigit:]]+'[^'])/'\\\x\\1/g" \
+-e '1s/^/BEGIN;\n/' \
+-e '$aCOMMIT;' /path/to/dump.sql
 ```
 
-## Generate New Schema
+### To MySQL
 
-Modify your `database_url` in `lldap_config.toml` (or `LLDAP_DATABASE_URL` in the env) to point to your new database. Restart LLDAP and check the logs to ensure there were no errors connecting and creating the tables. After that, stop LLDAP. Now we can import our original data!
+MySQL mostly cooperates, but it gets some errors if you don't escape the `groups` table. Run the
+following command to wrap all table names in backticks for good measure, and wrap the inserts in
+a transaction:
+
+```
+sed -i -r -e 's/^INSERT INTO ([a-zA-Z0-9_]+) /INSERT INTO `\1` /' \
```

Comment:
What does this line do? Can you break it down for me?

Reply:
yep! it goes from 
```
INSERT INTO groups VALUES
```
to
```
INSERT INTO `groups` VALUE
```

I think `groups` is a special word in mysql, it threw syntax errors for JUST the groups table when not escaped like that. It's listed as a [reserved keyword](https://dev.mysql.com/doc/refman/8.0/en/keywords.html).

This line generalizes it, but i could easily JUST update the `groups` insert statements

Reply:
oh, yeah, the `sed` part I had, it was more about the couple of lines below, sorry.

Reply:
Oh, the next one adds START TRANSACTION; to the beginning, and the one after adds COMMIT; to the end